sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala

```diff
@@ -412,6 +412,11 @@ case class CatalogRelation(
   assert(tableMeta.partitionSchema.sameType(partitionCols.toStructType))
   assert(tableMeta.dataSchema.sameType(dataCols.toStructType))
 
+  // If table schema is empty, Spark will infer the schema at runtime, so we should mark this
+  // relation as unresolved and wait it to be replaced by relation with actual schema, before
+  // resolving parent plans.
+  override lazy val resolved: Boolean = tableMeta.schema.nonEmpty
```

Comment:
I think the schema is generated by their programs. That is why it contains zero column.